Improved IPTP by fitting a cubic spline

rbbg opened the pull request:
As discussed in #160. I have used a GPL licensed spline fitting library ([this one](https://github.com/ttk592/spline)) to improve the performance of the iterative parabolic time parameterization algorithm that is used by MoveIt!. Unfortunately, the acceleration limits are still not satisfied completely, but it appears to be a step up from the previous approach.

Problems:

- The license. I am not too well versed in the what/why/how of open source licensing but I don't think GPL and BSD are compatible. Perhaps someone more knowledgeable on this front can offer some insight? If anyone is aware of a BSD licensed spline library that'd be great.

- The acceleration boundary conditions are not 0. This causes the waves you see between the first and second waypoints in the following figure:
![splinefit](https://cloud.githubusercontent.com/assets/8699378/21184312/33698338-c20c-11e6-99ab-a99bc47166d9.png)

davetcoleman commented:
I have emailed the Spline author @ttk592 asking if he would be ok changing the license to BSD.

rbbg commented:
That'd be pretty neat.

I am still pondering the issue with the first waypoint. I'd really like to figure out a way to make it work as it would mean you actually know what your maximum velocities and accelerations are, and you could use that information to perform a parameterization that is _guaranteed_ to not violate the limits on the controller side.

v4hn commented:
@rbbg Thanks for providing the pull-request!

Do you already got permission to change the license? If you can't point to some text by the author that allows you to do that, please undo it in the pull-request again! I know it's frustrating, but software licensing is nothing to toy with in our case. We definitely do *not* want people not to use the project because of murky licensing practice...

I do not yet understand where your "first waypoint" issue stems from.
Would it help to add another (zero-acceleration) waypoint in front of the trajectory before fitting?

davetcoleman commented on moveit_core/trajectory_processing/src/iterative_time_parameterization.cpp:
```diff
@@ -211,81 +208,26 @@ void updateTrajectory(robot_trajectory::RobotTrajectory &rob_trajectory, const s
 
-  // Accelerations
-  for (int i = 0; i < num_points; ++i)
+  // Spline fitting using GPL lib
```
maybe not call this "GPL lib" but something like "cubic spline library"

davetcoleman commented on moveit_core/trajectory_processing/src/iterative_time_parameterization.cpp:
```diff
@@ -35,6 +35,7 @@
 /* Author: Ken Anderson */
```
Given your approach is still experimental, I think we should not change Ken Anderson's itarative_time_parameterization version but rather add a secondary approach that can be optionally changed to (and documented). So I think you should copy this file and call it something different. You will also need to copy the [planning request adaptor plugin](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/planning/planning_request_adapter_plugins/src/add_time_parameterization.cpp) and name it something different. This will ensure we don't change the behavior of currently deployed systems.

rbbg replied:
Yeah, I figured this to be more of an initial testing version, but if it can actually be merged I will move it to another file and give the plugin another name.

ttk592 commented:
The license change to BSDv3 in this commit is fine by me.

rbbg commented:
As for a quick comparison of the performance: a 7dof manipulator planning around some obstacles(usually around 20-30 states):

They both run in 1-2ms and the Spline approach is _usually_ around 10-20% slower, so 1.7ms for Spline, 1.5ms for the original approach. In both cases, the iterative procedure to determine the time differences between waypoints takes the most time (The original heuristic setting of the velocities and accelerations takes <100us)

rbbg commented:
I fixed the problem with the first waypoint by prefixing a zero acceleration waypoint 1ns before the start waypoint, _after_ fitting the spline. 

One problem remains, with the current starting timestamps, the first waypoint often gets dropped by the controller because it occurs in the past, when this happens, the controller will again apply the quintic interpolation and you get the 'wave' as seen in the figure above. I've added a 50ms offset for now to ensure this doesn't happen, but that's not a very good solution.

Currently the trajectories are looking like the following figure (acc limit = 1.0). 

![figure_3](https://cloud.githubusercontent.com/assets/8699378/21257072/2a2d97e4-c376-11e6-826c-6cb921c6d456.png)

Still not satisfying the joint limits, but at least the acceleration is limited by the values that are in the waypoints.

I have named the plugin `SplineParameterization` and it can be added to the pipeline by `default_planner_request_adapters/AddSplineParameterization`.

davetcoleman commented on moveit_core/trajectory_processing/include/moveit/trajectory_processing/spline_parameterization.h:
```diff
@@ -0,0 +1,76 @@
+ */
+
+#ifndef MOVEIT_TRAJECTORY_PROCESSING_SPLINE_SMOOTHER_
+#define MOVEIT_TRAJECTORY_PROCESSING_SPLINE_SMOOTHER_
```
Should be SPLINE_PARAMETERIZATION_

rbbg commented:
I have made the requested changes, anything else required? 

To clarify: This only _adds_ a parameterization method, so the existing IPTP method remains unchanged.

davetcoleman commented:
@rbbg can you please summarize the benefits discussed in this thread and document them in the Parameterization Tutorial I started? http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/time_parameterization_tutorial.html

Thanks!
